anndata/_core/sparse_dataset.py

```diff
@@ -343,7 +343,29 @@ def __repr__(self) -> str:
     def __getitem__(self, index: Index | tuple[()]) -> float | ss.spmatrix:
         row, col = self._normalize_index(index)
         mtx = self._to_backed()
-        sub = mtx[row, col]
+
+        # see https://github.com/scverse/anndata/issues/1224 for special handling of boolean masks
+        # scipy internally converts boolean masks to integer indices so this needs to occur before access to `mtx`.
+        def make_slices(mask):
+            return np.ma.extras._ezclump(mask)
+
+        def mean_slice_length(slices):
+            return floor((slices[-1].stop - slices[0].start) / len(slices))
+
+        if (
+            self.format == "csr"
+            and np.array(row).dtype == bool
+            and mean_slice_length(make_slices(row)) > 7
+        ):
```

Comment:
I will try to come up a less repetitive way of doing this, especially since `make_slices` might not be so cheap in the future with larger datasets or something....

Reply:
Any luck on this?

Reply:
Yeah, I think it may have gotten more complicated...

I was thinking more:

* Don't include definitions for `make_slices` and `mean_slice_length` in here
* Keep the flow control logic, but make it like:


```python
if (
    self.format == "csr"
    and np.array(row).dtype == bool
):
    sub = sparse.csr_matrix(_mask_major_axis(mtx, row), shape=(mask.sum(), mtx.shape[1]))[:, col]
elif (
    self.format == "csc"
    and np.array(col).dtype == bool
):
    sub = sparse.csc_matrix(_mask_major_axis(mtx, col), shape=(mtx.shape[0], mask.sum()))[row, :]
```

Where

```python
def _mask_major_axis(mtx: sparse.spmatrix, mask: np.ndarray) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
    slices = make_slices(mask)

    if mean_slice_length(slices) <= 7:
        return get_compressed_vectors(mtx.data, mtx.indices, mtx.indptr, np.where(mask)[0])
    else:
        return get_compressed_vector_for_slices(mtx.data, mtx.indices, mtx.indptr, slices)
```